### DISCUSSION TITLE
Why stuck on WinUI 2?

### DISCUSSION THREAD

**Discussion by Lakshrp112** [microsoft/terminal#17127] [Q&A] [answered]
I recently came across a statement indicating, "and we're pretty stuck on WinUI 2 indefinitely." I'm curious to understand why Windows Terminal is sticking to WinUI 2 despite being a first-party application. I've thoroughly searched GitHub issues, Microsoft blogs, and more, but haven't found a clear answer yet.

Despite being Microsoft's first-party terminal solution, positioned as a replacement for CMD and PowerShell, Windows Terminal has not migrated to its own solution, even though Microsoft is actively encouraging third-party developers to do so. This discrepancy raises questions about why Windows Terminal, as a flagship Microsoft application, is not taking advantage of newer frameworks like WinUI 3.

Additionally, I'd like to mention that I have no experience or knowledge about native Windows application development. However, I'm interested in understanding why Windows Terminal is sticking to WinUI 2, despite being a first-party application.

The migration to WinUI 3, facilitated by the "Windows App SDK" (formerly known as Project Reunion), would offer a seamless transition without the need for a full app rewrite, as claimed by the Windows App SDK itself. This move could enhance the user experience and enable the adoption of modern features such as proper implementation of Mica.

**Comment by zadjii-msft** [comment, marked as answer]
There's one key thing that prevents us from using WinUI 3. SwapchainPanels on top of acrylic (and other XAML content). With the way the terminal is currently architected, each pane is a swapchainpanel with a transparent background. This lets us layer the swapchain _on top_ of XAML content. Namely, we use that so the background of a pane can be acrylic, or an image (namely: a gif). 

Alas, undocked swapchains can't have transparent backgrounds. So that would mean no acrylic, and no background images. And people love their backgrounds. We've got open dependencies filed to get that sorted out, but it'll probably be a while before those get resolved. 

Other than that, the transition isn't that painful. We've even had someone internally prototype the conversion, so we know it can be done. We're also _way off the rails_ in the way we're using a swapchainpanel - we always have been. This isn't something I'd expect other developers to run into in a major way.

**Reply by cnbluefire** [reply]
https://github.com/user-attachments/assets/d85d6002-4502-41b2-b2c6-57cdd2b9f809

[HotLyric](https://github.com/cnbluefire/HotLyric)

I created this application using WinUI 3 and Win2D. The buttons utilize WinUI 3 controls; the lyric animations in the middle are done using Win2D's CanvasAnimatedControl, which internally uses WinUI 3's SwapChainPanel; for the background, a custom acrylic brush was created using Windows.UI.Composition.

(translate by chatgpt)

**Reply by roke-julian-lockhart** [reply]
> https://github.com/microsoft/terminal/discussions/17127#discussioncomment-9223939

@zadjii-msft, I would rather you updated to WinUI3 and removed the ability to set a custom backdrop.

**Reply by zadjii-msft** [reply]
@RokeJulianLockhart What features from WinUI3 would you be looking for in particular? There'd have to be a compelling reason to do the non-trivial engineering work to move to WinUI 3

**Reply by roke-julian-lockhart** [reply]
> https://github.com/microsoft/terminal/discussions/17127#discussioncomment-10180493

@zadjii-msft, ah, nothing in particular. I just have no interest in that feature - it seems strange to withhold such an important architectural improvement for what seems like at best a gimmick.

**Reply by zadjii-msft** [reply]
People really love acrylic terminals 🤷 I'm not keen to remove a much beloved feature without a plan to replace it. Conversations are ongoing internally with the WinUI & platform folks to come up with a solution here. Something that benefits us will end up benefitting the whole of the platform.

**Reply by DHowett** [reply]
> important architectural improvement

WinUI 3 is only _arguably_ an improvement, and only because WinUI 2 has had bug fixes withheld to make WinUI 3 more compelling.

When it is possible to onboard to WinUI 3 and not take a hit in disk footprint, commit charge and startup time... then it will be an improvement. 🙂

**Reply by roke-julian-lockhart** [reply]
> https://github.com/microsoft/terminal/discussions/17127#discussioncomment-10182139

@DHowett, I estimate that those fixes you refer to were breaking? If so, that makes sense, right? Otherwise, I can't think of what else justifies a new version.
